model/docs/extending_pyrenew.qmd

```diff
@@ -0,0 +1,256 @@
+
+```python
+from pyrenew.metaclass import RandomVariable
+
+class MyNormal(RandomVariable):
+    def __init__(self, loc, scale):
+        self.validate(scale)
+        self.loc = loc
+        self.scale = scale
+        return None
+
+    @staticmethod
+    def validate(self):
+        if self.scale <= 0:
+            raise ValueError("Scale must be positive")
+        return None
+
+    def sample(self, **kwargs):
+        return (dist.Normal(loc=self.loc, scale=self.scale),)
+```
+
+The `@staticmethod` decorator exposes the `validate` function to be used outside the class. Next, we show how to build a more complex `RandomVariable` class; the `InfectionsWithFeedback` class.
+
+### The `InfectionsWithFeedback` class
+
+Although returning namedtuples are not strictly required, they are recommended as they make the code more readable. The following code-chunk shows how to create a named tuple for the `InfectionsWithFeedback` class:
+
+```{python}
+#| label: data-class
+from collections import namedtuple
+
+# Creating a tuple to store the output
+InfFeedbackSample = namedtuple(
+  typename='InfFeedbackSample',
+  field_names=['infections', 'rt'],
+  defaults=(None, None)
+)
+```
+
+The next step is to create the actual class. The bulk of its implementation lies in the function `pyrenew.latent.compute_infections_from_rt_with_feedback()`. We will also use the `pyrenew.datautils.pad_x_to_match_y()` function to ensure the passed vectors match their lengths. The following code-chunk shows most of the implementation of the `InfectionsWithFeedback` class:
+
+```{python}
+#| label: new-model-def
+#| code-line-numbers: true
+# Creating the class
+from pyrenew.metaclass import RandomVariable
+from pyrenew.latent import compute_infections_from_rt_with_feedback
+from pyrenew import datautils as du
+from jax.typing import ArrayLike
+
+class InfFeedback(RandomVariable):
+    """Latent infections"""
+
+    def __init__(
+        self,
+        infection_feedback_strength: RandomVariable,
+        infection_feedback_pmf: RandomVariable,
+        infections_mean_varname: str = "latent_infections",
+    ) -> None:
+        """Constructor"""
+
+        self.infection_feedback_strength = infection_feedback_strength
+        self.infection_feedback_pmf = infection_feedback_pmf
+        self.infections_mean_varname = infections_mean_varname
+
+        return None
+
+    def validate(self):
+        """
+        Generally, this method should be more meaningful, but we will skip it for now
+        """
+        return None
+
+    def sample(
+        self,
+        Rt: ArrayLike,
+        I0: ArrayLike,
+        gen_int: ArrayLike,
+        **kwargs,
+    ) -> tuple:
+        """Sample infections with feedback"""
+
+        # Baseline infections
+        I0_vec = du.pad_x_to_match_y(x=I0, y=Rt)
+        I0_vec = jnp.flip(I0_vec)
```

Comment:
Would prefer to call it `I0_vec_rev`, or, ideally, just pad it correctly in the first place.

Reply:
Addressed in 5eb31cc